client/rest-high-level/build.gradle

```diff
@@ -106,33 +108,49 @@ File nodeTrustStore = file("./testnode.jks")
 File pkiTrustCert = file("./src/test/resources/org/elasticsearch/client/security/delegate_pki/testRootCA.crt")
 
 integTest.runner {
+  systemProperty 'tests.rest.async', 'false'
   systemProperty 'tests.rest.cluster.username', System.getProperty('tests.rest.cluster.username', 'test_user')
   systemProperty 'tests.rest.cluster.password', System.getProperty('tests.rest.cluster.password', 'test-password')
 }
 
-testClusters.integTest {
-  testDistribution = 'DEFAULT'
-  systemProperty 'es.scripting.update.ctx_in_params', 'false'
-  setting 'reindex.remote.whitelist', '[ "[::1]:*", "127.0.0.1:*" ]'
-  setting 'xpack.license.self_generated.type', 'trial'
-  setting 'xpack.security.enabled', 'true'
-  setting 'xpack.security.authc.token.enabled', 'true'
-  setting 'xpack.security.authc.api_key.enabled', 'true'
-  // Truststore settings are not used since TLS is not enabled. Included for testing the get certificates API
-  setting 'xpack.security.http.ssl.certificate_authorities', 'testnode.crt'
-  setting 'xpack.security.transport.ssl.truststore.path', 'testnode.jks'
-  setting 'xpack.security.authc.realms.file.default_file.order', '0'
-  setting 'xpack.security.authc.realms.native.default_native.order', '1'
-  setting 'xpack.security.authc.realms.pki.pki1.order', '2'
-  setting 'xpack.security.authc.realms.pki.pki1.certificate_authorities', '[ "testRootCA.crt" ]'
-  setting 'xpack.security.authc.realms.pki.pki1.delegation.enabled', 'true'
-
-  setting 'indices.lifecycle.poll_interval', '1000ms'
-  keystore 'xpack.security.transport.ssl.truststore.secure_password', 'testnode'
-  user username:  System.getProperty('tests.rest.cluster.username', 'test_user'),
-          password:  System.getProperty('tests.rest.cluster.password', 'test-password')
-
-  extraConfigFile nodeCert.name, nodeCert
-  extraConfigFile nodeTrustStore.name, nodeTrustStore
-  extraConfigFile pkiTrustCert.name, pkiTrustCert
+// TODO: we can't use task avoidance here because RestIntegTestTask does the testcluster creation
```

Comment:
The use of `RestTestRunnerTask` would be preferable here. 
You __can__ have two tasks use the same cluster ( even trough right now it disabled the build cache ).
Just need to add `useCluster testClusters.integTest` to the task, and it can use task avoidance too.
No need for the `asyncIntegTest` as it won't really be used at the same time. 
Assuming cleanup works as it should, we should be able to run the tests against the same cluster. 

Another approach is to add sub-projects so the tests can run in parallel. 
The downside is that we don't have a generic solution to wire the tests up to the sub-projects ( even trough we have examples elsewhere in the code ) 

@mark-vieira might also have an opinion on this.

Reply:
Id love some better advice here, especially if it does not double the build time for this heh

Reply:
If time is the primary concern, then sub-projects are the way to go. 
That will matter when running a small number of tasks, like running the tests for the client only so there are resources to spare on the machine,  it won't matter as much for large runs that run everything in CI where there are plenty of projects to compete for the available resources.

Reply:
I don't think we should do subprojects until we have a simple and clear way to set this up via a gradle plugin. There is complexity involved in properly transfering the test classes via a jar, extracting them, then convincing the test runner in the subproject to use the extracted class files. Another task here is fine (and I think another cluster is fine too, we don't need to share, since we link the main directories there is minimal that is copied, and the setup here is minimal since it is not a rest test). It matches what we did for `transport-netty4` with pooled/unpooled tests.

Reply:
I think we should use two clusters and two tasks. Having multiple tasks targeting the same cluster disables caching. Having multiple clusters that are technically identical doesn't actually have much in the way of overhead due to the way we set these up with hardlinks but the benefit of being able to cache these tests is potentially significant.

I don't completely understand this comment:
```
TODO: we can't use task avoidance here because RestIntegTestTask does the testcluster creation
```

This is technically correct, creating a `RestIntegTestTask` creates the corresponding cluster but this doesn't affect build avoidance in any way. Everything is still wired up in a cacheable way.

Also, -1 to multiple subprojects. These test executions are identical in every way except for a single system property set on the test executor itself. There's absolutely no reason to create subprojects except for the fact that would facilitate running these suites in parallel which I don't think is very compelling given the way our builds are structured and already highly parallelizable.

Reply:
I do think that this argument is compelling. It's so unfortunate that there's no way to run test tasks in parallel for the same project! 
For someone developing the client who has to run these tests often it will make a big difference as it's likely to cut the time it takes in half.  I do agree it doesn't make a bug difference for larger runs in CI running tests for a large number of projects. 
I 100% agree that we should have a plugin if we are to go with the sub-project route, now is as good of a time as any. 
Just to be clear @hub-cap , I don't think there's anything here to prevent mergin the PR I'm just happy it triggered the discussion.

Reply:
> I do think that this argument is compelling. It's so unfortunate that there's no way to run test tasks in parallel for the same project!

This is better solved in Gradle core by moving test executing to the worker API. I know this is something they _want_ to do but not sure where it stacks up on the roadmap. I'm still highly annoyed that they got rid of `@ParallelizableTask` in Gradle 4.0 w/o a proper replacement.

Reply:
Thanks for the review friends, so from what I gather I can leave it as is (sans fixing the nits of course).

as for the comment @mark-vieira, it was copied from an earlier comment in some other code that did something similar, I just did not want to lose whatever reference to what was in the original authors head. But it sounds like I can just remove this comment in both places from your comment.